src/smc-webapp/course/handouts_panel.cjsx

```diff
@@ -0,0 +1,563 @@
+
+    copy_handout_to_all : (step, new_only) ->
+        @props.actions.copy_handout_to_all_students(@props.handout, new_only)
+
+    render_more_header : ->
+        <div>
+            <Button onClick={@open_handout_path}>
+                <Icon name="folder-open-o" /> Edit Handout
+            </Button>
+        </div>
+
+    render_handout_notes : ->
+        <Row key='note' style={styles.note}>
+            <Col xs=2>
+                <Tip title="Notes about this handout" tip="Record notes about this handout here. These notes are only visible to you, not to your students.  Put any instructions to students about handouts in a file in the directory that contains the handout.">
+                    Private Handout Notes<br /><span style={color:"#666"}></span>
+                </Tip>
+            </Col>
+            <Col xs=10>
+                <MarkdownInput
+                    rows          = 6
+                    placeholder   = 'Private notes about this handout (not visible to students)'
+                    default_value = {@props.handout.get('note')}
+                    on_save       = {(value)=>@props.actions.set_handout_note(@props.handout, value)}
+                />
+            </Col>
+        </Row>
+
+    render_copy_all : (status) ->
+        steps = STEPS()
+        for step in steps
+            if @state["copy_confirm_#{step}"]
+                @render_copy_confirm(step, status)
+
+    render_copy_confirm : (step, status) ->
+        <span key="copy_confirm_#{step}">
+            {@render_copy_confirm_to_all(step, status) if status[step]==0}
+            {@render_copy_confirm_to_all_or_new(step, status) if status[step]!=0}
+        </span>
+
+    render_copy_cancel : (step) ->
+        cancel = =>
+            @setState("copy_confirm_#{step}":false, "copy_confirm_all_#{step}":false, copy_confirm:false)
+        <Button key='cancel' onClick={cancel}>Cancel</Button>
+
+    copy_handout : (step, new_only) ->
+        # handout to all (non-deleted) students
+        switch step
+            when 'handout'
+                @props.actions.copy_handout_to_all_students(@props.handout, new_only)
+            else
+                console.log("BUG -- unknown step: #{step}")
+        @setState("copy_confirm_#{step}":false, "copy_confirm_all_#{step}":false, copy_confirm:false)
+
+    render_copy_confirm_to_all : (step, status) ->
+        n = status["not_#{step}"]
+        <Alert bsStyle='warning' key="#{step}_confirm_to_all", style={marginTop:'15px'}>
+            <div style={marginBottom:'15px'}>
+                {misc.capitalize(step_verb(step))} this handout {step_direction(step)} the {n} student{if n>1 then "s" else ""}{step_ready(step, n)}?
+            </div>
+            <ButtonToolbar>
+                <Button key='yes' bsStyle='primary' onClick={=>@copy_handout(step, false)} >Yes</Button>
+                {@render_copy_cancel(step)}
+            </ButtonToolbar>
+        </Alert>
```

Comment:
These small render methods could be moved off since it mostly shares with the same code in assignments. However, it is slightly different enough that I prefer to keep it as is until we have at least a third area where we have the same code use.